File: Inversion/inversion.py
from array import array

n = input()
MAXV = 50111
BLOCK = 777

ind = array('l', [])
for i in xrange(n+5):
    ind.append(0)
start = array('l', [])
last = array('l', [])
block_len = array('l', [])
sl = []
for i in xrange(411):
    start.append(0)
    last.append(0)
    block_len.append(0)

    tmp = array('l', [])
    for j in xrange(MAXV):
        tmp.append(0)
    sl.append(tmp)


def update(x, u, k):
    v = u
    while v <= MAXV:
        sl[x][v] += k
        v += v & (-v)


def get(x, u):
    res = 0
    v = u
    while v > 0:
        res += sl[x][v]
        v -= v & (-v)
    return res


a = array('l', [0] + map(int, raw_input().split()))
all_values = set(a)
value_map = {}
cnt = 0
for x in all_values:
    cnt += 1
    value_map[x] = cnt
for i in xrange(1, n+1):
    a[i] = value_map[a[i]]

res = 0

for i in xrange(n, 0, -1):
    update(0, a[i], 1)
    res += get(0, a[i] - 1)

nblock = 1
c = 0
for i in xrange(1, n+1, 1):
    if c == 0:
        start[nblock] = i
    ind[i] = nblock
    block_len[nblock] += 1
    c += 1
    if c == BLOCK:
        last[nblock] = i
        nblock += 1
        c = 0

last[nblock] = n
for i in xrange(1, n+1):
    update(ind[i], a[i], 1)

q = input()
for qq in xrange(q):
    u, val = map(int, raw_input().split())
    p = ind[u]

    # Block before u
    for i in xrange(1, p):
        res -= block_len[i] - get(i, a[u])
        res += block_len[i] - get(i, val)

    # Block after u
    for i in xrange(p+1, nblock+1):
        res -= get(i, a[u] - 1)
        res += get(i, val - 1)

    # Block containing u
    for i in xrange(start[p], u):
        if a[i] > a[u]:
            res -= 1
        if a[i] > val:
            res += 1

    for i in xrange(u+1, last[p]+1):
        if a[i] < a[u]:
            res -= 1
        if a[i] < val:
            res += 1

    # Update block containing u
    update(p, a[u], -1)
    a[u] = val
    update(p, val, +1)

    print res

File: Inversion/inversion.java
import java.io.*;
import java.util.*;
import java.math.*;

public class inversion {
    public static void main(String[] args) {
        (new Solver()).go();
    }
}

class Solver {
    int MN = 250111;
    int MAXV = 50111;
    int BLOCK = 777;

    int n, q, nblock;
    long res;
    int[] a = new int[MN];
    int[] ind = new int[MN];
    int[][] sl = new int[411][MAXV];
    int[] start = new int[411];
    int[] last = new int[411];
    int[] len = new int[411];

    void update(int x, int u, int k) {
        for(int v = u; v <= MAXV; v += v & (-v)) {
            sl[x][v] += k;
        }
    }

    int get(int x, int u) {
        int res = 0;
        for(int v = u; v > 0; v -= v & (-v)) {
            res += sl[x][v];
        }
        return res;
    }

    public void go() {
        InputStream inputStream = System.in;
        OutputStream outputStream = System.out;
        InputReader in = new InputReader(inputStream);
        PrintWriter out = new PrintWriter(outputStream);

        int n = in.nextInt();
        for(int i = 1; i <= n; ++i)
            a[i] = in.nextInt();

        for(int i = n; i >= 1; --i) {
            update(0, a[i], 1);
            res += get(0, a[i] - 1);
        }

        nblock = 1;
        int c = 0;
        for(int i = 1; i <= n; ++i) {
            if (c == 0) start[nblock] = i;
            ind[i] = nblock; len[nblock]++;
            ++c;

            if (c == BLOCK) {
                last[nblock] = i;
                ++nblock; c = 0;
            }
        }
        last[nblock] = n;
        for(int i = 1; i <= n; ++i)
            update(ind[i], a[i], 1);

        int q = in.nextInt();
        for(int qq = 0; qq < q; ++qq) {
            int u = in.nextInt(), val = in.nextInt();
            int p = ind[u];

            // block before u
            for(int i = 1; i < p; ++i) {
                res -= len[i] - get(i, a[u]);
                res += len[i] - get(i, val);
            }
            // block after u
            for(int i = p+1; i <= nblock; ++i) {
                res -= get(i, a[u] - 1);
                res += get(i, val - 1);
            }

            // block containing u
            for(int i = start[p]; i <= u-1; ++i) {
                if (a[i] > a[u]) --res;
                if (a[i] > val) ++res;
            }
            for(int i = u + 1; i <= last[p]; ++i) {
                if (a[i] < a[u]) --res;
                if (a[i] < val) ++res;
            }

            // update block u
            update(p, a[u], -1);
            a[u] = val;
            update(p, val, +1);

            System.out.println(res);
        }
    }
}

class InputReader {
    public BufferedReader reader;
    public StringTokenizer tokenizer;

    public InputReader(InputStream stream) {
        reader = new BufferedReader(new InputStreamReader(stream), 32768);
        tokenizer = null;
    }

    public String next() {
        while (tokenizer == null || !tokenizer.hasMoreTokens()) {
            try {
                tokenizer = new StringTokenizer(reader.readLine());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return tokenizer.nextToken();
    }

    public int nextInt() {
        return Integer.parseInt(next());
    }

}
